src/Builder/BuilderEncoder.php

```diff
@@ -4,323 +4,101 @@
-                    $result->{'$geometry'} = $val->{'$geometry'};
-                } else {
-                    $result->{'$geometry'} = $val;
-                }
-            } else {
-                $result->{'$' . $key} = $val;
-            }
+        if (! is_object($value)) {
+            return false;
         }
 
-        return $this->wrap($value, $result);
-    }
-
-    /**
-     * Get the unique property of the operator as value
-     */
-    private function encodeAsSingle(OperatorInterface $value): stdClass
-    {
-        foreach (get_object_vars($value) as $val) {
-            $result = $this->recursiveEncode($val);
-
-            return $this->wrap($value, $result);
-        }
-
-        throw new LogicException(sprintf('Class "%s" does not have a single property.', $value::class));
-    }
-
-    private function encodeCombinedFilter(CombinedFieldQuery $filter): stdClass
-    {
-        $result = new stdClass();
-        foreach ($filter->fieldQueries as $filter) {
-            $filter = $this->recursiveEncode($filter);
-            if (is_object($filter)) {
-                $filter = get_object_vars($filter);
-            } elseif (! is_array($filter)) {
-                throw new LogicException(sprintf('Query filters must an array or an object. Got "%s"', get_debug_type($filter)));
-            }
-
-            foreach ($filter as $key => $value) {
-                $result->{$key} = $value;
-            }
-        }
+        $encoder = $this->getEncoderFor($value);
 
-        return $result;
+        return $encoder !== null && $encoder->canEncode($value);
     }
 
-    /**
-     * Query objects are encoded by merging query operator with field path to filter operators in the same object.
-     */
-    private function encodeQueryObject(QueryObject $query): stdClass
+    public function encode(mixed $value): stdClass|array|string
     {
-        $result = new stdClass();
-        foreach ($query->queries as $key => $value) {
-            if ($value instanceof QueryInterface) {
-                // The sub-objects is merged into the main object, replacing duplicate keys
-                foreach (get_object_vars($this->recursiveEncode($value)) as $subKey => $subValue) {
-                    if (property_exists($result, (string) $subKey)) {
-                        throw new LogicException(sprintf('Duplicate key "%s" in query object', $subKey));
-                    }
+        $encoder = $this->getEncoderFor($value);
 
-                    $result->{$subKey} = $subValue;
-                }
-            } else {
-                if (property_exists($result, (string) $key)) {
-                    throw new LogicException(sprintf('Duplicate key "%s" in query object', $key));
-                }
-
-                $result->{$key} = $this->encodeIfSupported($value);
-            }
+        if (! $encoder || ! $encoder->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
```

Comment:
The `$encoder->canEncode($value)` part is already checked inside `$encoder->encode($value)`.

Reply:
Yes, however there is no guarantee that somebody will never call `encode` with an invalid value, as the contract does not require calling `canEncode` for a value beforehand.

We could solve this with the introduction of a `null` type for the `$value` argument in the `Encoder::encode` definition, but that would require us to always accept `null` values in any encoder, which might still necessitate calling `canEncode`. `void` as PHP's true bottom type is unfortunately not available as an argument type, so there's no proper typing way around this.